src/ts/api-service.ts

```diff
@@ -0,0 +1,116 @@
+        this.gridSize = numberOfRecords;    //Initialise the class with the available size of the grid
+
+        //Send AJAX request to server to retrieve the all the column names to display
+        $.ajax({
+            "url": this.url + "/columns",
+            "success": data => {
+                this.columnNames = JSON.parse(data);
+            }
+        });
+    }
+
+    recordCount() {
+        return new Promise((resolve, reject) => {
+            $.ajax({
+                "url": this.url + "/recordCount",
+                "success": data => {
+                    resolve(this.totalRecords = Number(data));
+                },
+                "error": (e) => {
+                    reject(console.log(e))  //Log the error for debugging purposes
+                }
+            });
+        })
+    }
+
+    getCurrentRecords() {
+        let toId: number;
+        //Calculate the "to" parameter for the records to collect
+        if (this.topRecordIndex + (this.gridSize - 1) > (this.totalRecords - 1)) {
+            toId = this.totalRecords - 1;
+            this.topRecordIndex = this.totalRecords - this.gridSize;
+        } else {
+            toId = this.topRecordIndex + (this.gridSize - 1);
+        }
+        return new Promise((resolve, reject) => {
+            $.ajax({
+                "url": this.url + "/records",
+                "data": {
+                    "from": this.topRecordIndex.toString(),
+                    "to": toId.toString()
+                },
+                "success": data => {
+                    resolve(this.dataRecords = JSON.parse(data))    //store result privately
+                },
+                "error": (e) => {
+                    reject(console.log(e))  //Log the error for debugging purposes
+                }
+            })
+        })
+    }
+
+    previous() {
+        //Check if there are records to the left of the top index of the grid before sending a request
+        if (this.topRecordIndex - (this.gridSize - 1) > -1) {
+            this.topRecordIndex -= this.gridSize;
+            return this.getCurrentRecords();
+        } else {
+            if (this.topRecordIndex === 0) {
+                return null;
+            } else {
+                this.topRecordIndex = 0;
+                return this.getCurrentRecords();
+            }
+        }
+    }
+
+    next() {
+        //Check if there are records to the right of the top index of the grid before sending a request
+        if (this.topRecordIndex + (this.gridSize - 1) < (this.totalRecords - 1)) {
+            this.topRecordIndex += this.gridSize;
+            return this.getCurrentRecords();
+        } else {
+            return null;
+        }
+    }
+
+    searchRecord(id: string) {
+        this.topRecordIndex = Number(id);   //The searched value will always be the top record's ID
+        return this.getCurrentRecords().then(() => {
+            for (let record of this.dataRecords) {
```

Comment:
I don't think you need the loop if the only operation in it is 
```
                if (record[0] == id) {
                    record.unshift('searched')
                }
```

just remove loop and do the single operation using this.dataRecords

Reply:
So I want to push the first column (record[0] is the ID at first) of the searched row with the keyword 'searched' so that I can identify it and style it.

The reason for looping is because I don't know the index of the 'searched' row as my index of the array is not the same as the searched ID. I only fetch the fraction of the total records that are to be displayed. So the ID and index are misaligned

Reply:
Ah, right